vm/src/stdlib/array.rs

```diff
@@ -0,0 +1,306 @@
+                        let slice = unsafe { std::slice::from_raw_parts(ptr, ptr_len) };
+                        v.extend_from_slice(slice);
+                    })*
+                }
+            }
+
+            fn tobytes(&self) -> Vec<u8> {
+                match self {
+                    $(ArrayContentType::$n(v) => {
+                        // safe because we're just reading memory as bytes
+                        let ptr = v.as_ptr() as *const u8;
+                        let ptr_len = v.len() * std::mem::size_of::<$t>();
+                        let slice = unsafe { std::slice::from_raw_parts(ptr, ptr_len) };
+                        slice.to_vec()
+                    })*
+                }
+            }
+
+            fn index(&self, x: PyObjectRef, vm: &VirtualMachine) -> PyResult<Option<usize>> {
+                match self {
+                    $(ArrayContentType::$n(v) => {
+                        let val = $t::try_from_object(vm, x)?;
+                        Ok(v.iter().position(|&a| a == val))
+                    })*
+                }
+            }
+
+            fn reverse(&mut self) {
+                match self {
+                    $(ArrayContentType::$n(v) => v.reverse(),)*
+                }
+            }
+
+            fn getitem(&self, i: usize, vm: &VirtualMachine) -> Option<PyResult> {
+                match self {
+                    $(ArrayContentType::$n(v) => v.get(i).map(|x| x.into_pyobject(vm)),)*
+                }
+            }
+        }
+    };
+}
+
+def_array_enum!(
+    (SignedByte, i8, 'b'),
+    (UnsignedByte, u8, 'B'),
+    // TODO: support unicode char
+    (SignedShort, i16, 'h'),
+    (UnsignedShort, u16, 'H'),
+    (SignedInt, i16, 'i'),
+    (UnsignedInt, u16, 'I'),
+    (SignedLong, i32, 'l'),
+    (UnsignedLong, u32, 'L'),
+    (SignedLongLong, i64, 'q'),
+    (UnsignedLongLong, u64, 'Q'),
+    (Float, f32, 'f'),
+    (Double, f64, 'd'),
+);
+
+#[pyclass]
+#[derive(Debug)]
+pub struct PyArray {
+    array: RefCell<ArrayContentType>,
+}
+pub type PyArrayRef = PyRef<PyArray>;
+
+impl PyValue for PyArray {
+    fn class(vm: &VirtualMachine) -> PyClassRef {
+        vm.class("array", "array")
+    }
+}
+
+#[pyimpl]
+impl PyArray {
+    #[pymethod(name = "__new__")]
+    fn new(
+        cls: PyClassRef,
+        spec: PyStringRef,
+        init: OptionalArg<PyIterable>,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyArrayRef> {
+        let spec = match spec.as_str().len() {
+            1 => spec.as_str().chars().next().unwrap(),
+            _ => {
+                return Err(vm.new_type_error(
+                    "array() argument 1 must be a unicode character, not str".to_owned(),
+                ))
+            }
+        };
+        let array = ArrayContentType::from_char(spec).ok_or_else(|| {
+            vm.new_value_error(
+                "bad typecode (must be b, B, u, h, H, i, I, l, L, q, Q, f or d)".to_owned(),
```

Comment:
It might be nice to generate this error message in the `from_char` method, and let it return this error message and use `map_err`.